peps/pep-0793.rst

```diff
@@ -0,0 +1,476 @@
+export hook can use a slot with a new ID, ``Py_mod_token``, to set the token.
+
+Modules created from a ``PyModuleDef`` will have the token set to that
+definition. An explicit ``Py_mod_token`` slot will we rejected for these.
+(This allows implementations to share storage for the token and def.)
+
+For modules created via the new export hook, the token
+will be set to the address of the slots array by default.
+(This does **not** apply to modules created by ``PyModule_FromSlotsAndSpec``,
+as that function's input might not outlive the module.)
+
+The token will not be set for non-``PyModuleType`` instances.
+
+A ``PyModule_GetToken`` function will be added to get the token.
+Since the result may be ``NULL``, it will be passed via a pointer; the function
+will return 0 on success and -1 on failure:
+
+.. code-block:: c
+
+    int PyModule_GetToken(PyObject *, void **token_p)
+
+A new ``PyType_GetModuleByToken`` function will be added, with a signature
+like ``PyType_GetModuleByDef`` but a ``void *token`` argument,
+and the same behaviour except matching tokens, rather than only defs.
+
+
+New slots
+---------
+
+For each field of the ``PyModuleDef`` struct, except ones from
+``PyModuleDef_HEAD_INIT``, a new slot ID will be provided: ``Py_mod_name``,
+``Py_mod_doc``, ``Py_mod_clear``, etc.
+Slots related to the module state rather than the module object will
+use a ``Py_mod_state_`` prefix.
+See :ref:`pep793-api-summary` for a full list.
+
+All new slots -- these and ``Py_tp_token`` discussed above -- may not be
+repeated in the slots array, and may not be used in a
+``PyModuleDef.m_slots`` array.
+They may not have a ``NULL`` value (instead, the slot can be omitted entirely).
+
+Note that currently, for modules created from a *spec* (that is, using
+``PyModule_FromDefAndSpec``), the ``PyModuleDef.m_name`` member is ignored
+and the name from the spec is used instead.
+All API proposed in this document creates modules from a *spec*, and it will
+ignore ``Py_mod_name`` in the same way.
+The slot will be optional, but extensions will be strongly encouraged to
+include it for the benefit of future APIs, external tooling,
+and debugging/introspection.
+
+
+Bits & Pieces
+-------------
+
+A ``PyMODEXPORT_FUNC`` macro will be added, similar to the ``PyMODINIT_FUNC``
+macro but with ``PyModuleDef_Slot *`` as the return type.
+
+A ``PyModule_GetStateSize`` function will be added to retrieve the size set
+by ``Py_mod_state_size`` or ``PyModuleDef.m_size``.
+Since the result may be -1 (for single-phase-init modules), it will be output
+via a pointer; the function will return 0 on success and -1 on failure:
+
+.. code-block:: c
+
+    int PyModule_GetStateSize(PyObject *, Py_ssize_t *result);
+
+
+.. _pep793-api-summary:
+
+New API summary
+---------------
+
+The following functions will be added:
+
+.. code-block:: c
+
+    PyObject *PyModule_FromSlotsAndSpec(PyModuleDef_Slot *, PyObject *spec)
```

Comment:
The export hook is part of the API, but the functions are defined in extension modules. Worth including?
```suggestion
    PyModuleDef_Slot *PyModExport_<NAME>(PyObject *spec);
    PyObject *PyModule_FromSlotsAndSpec(PyModuleDef_Slot *, PyObject *spec)
```